[Feature] Support revocation of single certificates (EXPOSUREAPP-12455)

BMItr opened the pull request:
This PR addresses [EXPOSUREAPP-12455](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-12455).

Revocation of single certificates story.

How to test:
- Scan in a valid VC (MockCloud Environment)
- use CLI tool to revoke a certficate (newest version needed):
 _prompts for entering the barcode data, similar to `cwa dcc verify`
 then prompts for revoking or granting the DCC repeatedly_:
_cwa dcc rl -e mock-cloud --data -i_

- update the revocation list inside the dev menu or wait 24 h
- check if the certificate is revoked
- check if texts and UI matches to figma
- check if acceptance criteria are fulfilled

BMItr committed:
nothing

BMItr committed:
Support revocation skeleton (EXPOSUREAPP-12574) (#5038)

* Added RevocationModule

* Added models

* Added api and server

* Added RevocationEntryCoordinates

* Added update

* Added storage and repo

* Added RevocationCheck

* Added RevocationException

* Update RevocationUpdater.kt

* Adjusted to do

* Added RevocationCalculation

* little clean

* fix ident

* kt

* Update WorkerBinderTest.kt

* satisfy CI - hopefully

* satisfy CI

* missing space

* tests

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/revocation/RevocationModule.kt:
```diff
@@ -0,0 +1,45 @@
+import javax.inject.Singleton
+
+@Module
+object RevocationModule {
```
Maybe we should be specific here `DccRevocation...`?

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/revocation/RevocationModule.kt:
```diff
@@ -0,0 +1,45 @@
+        @AppContext context: Context,
+        @AppScope scope: CoroutineScope,
+        dispatcherProvider: DispatcherProvider
+    ): DataStore<Preferences> = PreferenceDataStoreFactory.create(
```
that is because you want to use Protobuff here or ?

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/revocation/calculation/RevocationCalculation.kt:
```diff
@@ -0,0 +1,40 @@
+import javax.inject.Inject
+
+@Suppress("FunctionOnlyReturningConstant")
+class RevocationCalculation @Inject constructor() {
```
Dcc..

mtwalli reviewed (commented):
Please add `Dcc` prefix to be more specific

mtwalli committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

mtwalli committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

mtwalli committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

mtwalli committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

BMItr committed:
Revocation calculations (EXPOSUREAPP-12553) (#5060)

* Added RevocationModule

* Added models

* Added api and server

* Added RevocationEntryCoordinates

* Added update

* Added storage and repo

* Added RevocationCheck

* Added RevocationException

* Update RevocationUpdater.kt

* Adjusted to do

* Added RevocationCalculation

* little clean

* fix ident

* kt

* Update WorkerBinderTest.kt

* satisfy CI - hopefully

* satisfy CI

* missing space

* tests

* Implemented calculateRevocationEntryTypeUCI

* Added RevocationCalculationTest

* Implemented calculateRevocationEntryTypeCOUNTRYCODEUCI

* Implemented calculateRevocationEntryTypeSIGNATURE

BMItr committed:
Revocation calculations adjustments (EXPOSUREAPP-12553) (#5067)

* Bigger than 1

* Use ParameterizedTest

mtwalli committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

BMItr committed:
Revocation - data structures & error codes (EXPOSUREAPP-12552) (#5064)

* Implemented RevocationKidList

* Implemented RevocationEntryCoordinates

* Added error codes and messages

* Implemented RevocationKidTypeIndex

* Implemented RevocationChunk

* Extracted RevocationHashType

* Added CachedRevocationKidTypeIndex

* Added CachedRevocationChunk

* Use ByteString

* Get RevocationHashType from byteString

* No persistence

* Return `CachedRevocationChunk`

Co-authored-by: Mohamed Metwalli <mohamed.metwalli@sap.com>

mtwalli committed:
Add revoked state (EXPOSUREAPP-12558) (#5058)

* Add new state

* Generate QrCode for revoked state

* Simplify

* Consider revoked in blocked state

* Doc

* state property

* Fix

* use `set`

* Map to ccl state

* Update `equals`

* add TODO

* REVOKED Cert text

* Fix merge issues

* qrcodeHash

* Refactoring

* Fix typo / compile errors

* Text changes

* Remove expiration

* Notifying states

* Fix test certificates state

* Update CertificateStateHelper.kt

* More TC state fixes

* Lint

BMItr committed:
Revocation parser (EXPOSUREAPP-12747) (#5076)

* Created RevocationParser

* Added RevocationParserTest

* added explanations to EntryCoord types

mtwalli committed:
Use coordinates as ID (#5079)

mtwalli committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

mtwalli committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

mtwalli committed:
Revocation check (EXPOSUREAPP-12557) (#5073)

* Add new state

* Generate QrCode for revoked state

* Simplify

* Consider revoked in blocked state

* Doc

* state property

* Fix

* use `set`

* Map to ccl state

* Update `equals`

* add TODO

* REVOKED Cert text

* Fix merge issues

* qrcodeHash

* Refactoring

* Fix typo / compile errors

* Text changes

* Remove expiration

* Notifying states

* Fix test certificates state

* Update CertificateStateHelper.kt

* More TC state fixes

* Lint

* Refactoring

* Checker logic

* More refactoring

* no usage

* message

* Calculate from `DccData`

* Update DccRevocationChecker.kt

* Lint

* Remove redundant

* kid

* Check against cached hashes

* Polish

* Lint

* Lint

* Fix compile error

* Test

* Create DccRevocationCheckerTest.kt

* Do check

* Testing...

* Update DccValidityMeasuresObserver.kt

mtwalli committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

BMItr committed:
Revocation Api (EXPOSUREAPP-12554) (#5080)

* Created RevocationParser

* Added RevocationParserTest

* added explanations to EntryCoord types

* Implemented RevocationApi

* Implemented RevocationServer

* Shorter param name

* - Implemented API creation, Added RevocationCache

* Adjusted RevocationServer

* Added RevocationRepository to DataReset

* Added RevocationServerTest

* Update HashExtensionsTest.kt

* clean

* Pay technical debt 💸

https://sonarcloud.io/project/issues?id=corona-warn-app_cwa-app-android&branch=feature%2F12555-revocation-storage-repo&resolved=false&types=CODE_SMELL

Co-authored-by: Mohamed Metwalli <mohamed.metwalli@sap.com>

mtwalli committed:
Update scheduler (EXPOSUREAPP-12556) (#5086)

* Schedule update

* Certificate addition

* Trigger on new registration

* Handle errors

* Comment

* Test menu

* update

* Reset Revocation settings

* Add unit tests

* Add unit test

* Create RevocationUpdateSchedulerTest.kt

* Lint

BMItr committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

BMItr committed:
Update CertificateStateHelper.kt

BMItr committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

BMItr committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

BMItr committed:
Revocation of single certs - storage, repo, updateservice (EXPOSUREAPP-12555, 12826) (#5104)

* Created RevocationParser

* Added RevocationParserTest

* added explanations to EntryCoord types

* Implemented RevocationApi

* Implemented RevocationServer

* Shorter param name

* - Implemented API creation, Added RevocationCache

* Adjusted RevocationServer

* Added RevocationRepository to DataReset

* Added RevocationServerTest

* Update HashExtensionsTest.kt

* clean

* Implemented RevocationStorage

* Added RevocationReset

* Create RevocationResetTest.kt

* Clear data store and cache directly

* Use data store directly

* repaired some conflicts from merge

* Adjusted clearing

* Added RevocationRepositoryTest

* Added RevocationUpdateService (WIP)

* Implementing RevocationUpdateService (WIP)

* Fixed disappearing RevocationEntryCoordinates

* Use set

* Log less, use coordinates directly

* Moved groupByNotNull

* Clean up and better logging

* some comments & cleanUp

* clean

* ktlint

* Update RevocationListUpdaterTest.kt

* Update uri to newest version

* clean

* Update RevocationApi.kt

* Update RevocationApi.kt

* Update RevocationApi.kt

* Update RevocationApi.kt

Co-authored-by: Mohamed Metwalli <mohamed.metwalli@sap.com>

BMItr committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

BMItr committed:
Revocation - scheduler adjustment (EXPOSUREAPP-12881) (#5110)

* Added launchInLatest

* Each repo also returns all certificates

* - Removed all cert size - Adjusted RevocationListUpdaterTest

* Adjusted RecoveryCertificateRepositoryTest

* Adjusted RecoveryCertificateRepositoryTest 2

* Adjusted TestCertificateRepositoryTest

* Adjusted VaccinationCertificateRepositoryTest

* neat and tidy 🧹

* missing get

* - Use jackson - Added ByteStringSerialization for jackson

* clean

BMItr committed:
Merge branch 'release/2.22.x' into feature/12455-feature-Revocation-of-single-certs

KathrinLuecke reviewed (approved):
UA approved

BMItr commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/revocation/storage/RevocationRepository.kt:
```diff
@@ -0,0 +1,75 @@
+
+@VisibleForTesting
+internal val CACHED_REVOCATION_CHUNKS_KEY by lazy {
+    stringPreferencesKey("RevocationStorage.revocationListJson")
```
No, it does not need to be lazy. But it can be lazy and has the benefit to get created off the main thread, even so it is not that costly to create.

BMItr commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/revocation/update/RevocationUpdateScheduler.kt:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private fun buildWorkRequest(): PeriodicWorkRequest {
+        return PeriodicWorkRequestBuilder<RevocationListUpdateWorker>(
```
will add that

mtwalli replied:
PLEASE don't use NETWORK constraints, There was a so long delay for CCL updates because of that and we had to remove it again :/

jurajkusnier reviewed (commented):
Revocation of RC/VC tested successfully on Pixel 4 (Android 12)

BMItr commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/expiration/DccValidityStateNotificationService.kt:
```diff
@@ -0,0 +1,112 @@
+    suspend fun showNotificationIfStateChanged(ignoreLastCheck: Boolean = false) = mutex.withLock {
+        Timber.tag(TAG).v("showNotificationIfStateChanged(ignoreLastCheck=%s)", ignoreLastCheck)
+        val lastCheck = covidCertificateSettings.lastDccStateBackgroundCheck.value
+        val timeUpdateRequired = lastCheck.toLocalDateUtc() == timeStamper.nowUTC.toLocalDateUtc()
```
Not my code, but i'll figure it out and adjust naming

mtwalli replied:
Same here the class was just renamed, but to answer you it check if the last `run` was in the same day and it is done that way to handle time travel too. DON't use time here please

BMItr commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/expiration/DccValidityStateNotificationService.kt:
```diff
@@ -0,0 +1,112 @@
+            return
+        }
+
+        val vacRecCerts = getCertificates()
```
I'll check that

mtwalli replied:
Keep in mind that Test certificates do NOT notify for expiration states

BMItr replied:
Will keep that name because getCertificates only loads rec and vac certs

BMItr commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/revocation/storage/RevocationRepository.kt:
```diff
@@ -0,0 +1,75 @@
+            prefs[CACHED_REVOCATION_CHUNKS_KEY]?.toCachedRevocationChunks() ?: emptyList()
+        }
+        .shareLatest(
+            tag = TAG,
```
We dont really need to log this. But keeping shareLatest is beneficial because it makes the flow distinctUntilChanged and conflated and in the future we may get new subscribers.

BMItr commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/revocation/update/RevocationListUpdater.kt:
```diff
@@ -0,0 +1,78 @@
+
+    init {
+        certificatesProvider.allCertificates
+            .distinctUntilChangedBy { it.size }
```
Not really but unfortunately it it not that easy to use distinctUntilChnages with new.size > old.size without edgecases..
unfortunately

BMItr commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/test/core/TestCertificateRepository.kt:
```diff
@@ -79,67 +82,72 @@ class TestCertificateRepository @Inject constructor(
-        }
+    val certificates: Flow<Set<TestCertificateWrapper>> = allCertificates
+        .map { it.certificates }
         .shareLatest(
```
No, will remove that logging

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/revocation/update/RevocationListUpdater.kt:
```diff
@@ -0,0 +1,78 @@
+        val lastExecution = revocationUpdateSettings.getLastUpdateTime() ?: return true
+
+        // update is needed if the last update was on a different day
+        return lastExecution.toLocalDateUtc() != now.toLocalDateUtc()
```
No, to work also in case of time travel

BMItr replied:
ok, will not adjust 😄

mtwalli commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/expiration/DccValidityStateNotificationService.kt:
```diff
@@ -0,0 +1,112 @@
+
+        val vacRecCerts = getCertificates()
+
+        vacRecCerts.notifyForState<Expired> {
```
The log statement yes ,but no notification unless the predicate if true.
BTW it has been always like that I just refactored it  to satisfy Sonar

BMItr replied:
Acutally it is kinda easy to understand and I like it that way
